File: JavaScript/6js 1/calc.js
/*
|--------------------------------------------------------------------------
| Reverse polish notation calculator
|--------------------------------------------------------------------------
|
*/

function calc(expr) {
   if (!expr || expr.trim() === "") return 0;
   const stack = [];
   const tokens = expr.trim().split(" ");
   for (let token of tokens) {
      if (!isNaN(token)) {
         stack.push(parseFloat(token));
      } else {
         const b = stack.pop();
         const a = stack.pop();
         switch (token) {
            case "+": stack.push(a + b); break;
            case "-": stack.push(a - b); break;
            case "*": stack.push(a * b); break;
            case "/": stack.push(a / b); break;
         }
      }
   }
   return stack.pop();
}


function calc(s) {
   var r = /(-?[\d\.]+) (-?[\d\.]+) ([-+\*\/])/;
   while (s != "" && r.test(s)) s = s.replace(r, (_, a, b, op) => eval(a + op + b));
   return +s.match(/-?[\d\.]+$/);
}


function calc(expr) {
   return expr
      .split(" ")
      .reduce(
         function (a, b) {
            return a.concat(
               b.match(/[0-9]/) ? parseFloat(b) : eval(a.splice(a.length - 2).join(b))
            );
         },
         [0]
      )
      .pop();
}


console.log(calc(""), 0);
console.log(calc("3"), 3);
console.log(calc("3.5"), 3.5);
console.log(calc("1 3 +"), 4);
console.log(calc("1 3 *"), 3);
console.log(calc("1 3 -"), -2);
console.log(calc("4 2 /"), 2);

File: Python/6py 1/calc.py
"""
|--------------------------------------------------------------------------
| Reverse polish notation calculator
|--------------------------------------------------------------------------
|
"""

from re import match
def calc(expr):
   stack, tokens = [0], expr.split(" ")
   for token in tokens:
      if match(r'\d', token):
         stack.append(float(token))
      else:
         operation = token.join(map(str, stack[-2:]))
         stack = stack[:-2]
         stack.append(eval(operation))
   return stack.pop()


def calc(expr):
   stack = []
   [stack.append(eval("{1}{2}{0}".format(stack.pop(), stack.pop(), e)) if e in ('+', '-', '/', '*') else e) for e in expr.split()]
   return float(stack.pop()) if stack else 0


def calc(expr):
   if len(expr) > 0:
      a,b = ['+','-','*','/'],[]
      for i in expr.split():
         if i not in a:
            b.append(i)
         else:
            n = int(b.pop())
            m = int(b.pop())
            if i == '+': c = m + n
            elif i == '-': c = m - n
            elif i == '*': c = m * n
            elif i == '/': c = m / n
            b.append(str(c))
      return eval(b[0])
   return 0


print(calc(""), 0)
print(calc("3"), 3)
print(calc("3.5"), 3.5)
print(calc("1 3 +"), 4)
print(calc("1 3 *"), 3)
print(calc("1 3 -"), -2)
print(calc("4 2 /"), 2)